src/librustc_typeck/collect.rs

```diff
@@ -1299,13 +1300,37 @@ fn trait_def_of_item<'a, 'tcx>(ccx: &CrateCtxt<'a, 'tcx>,
         err.emit();
     }
 
+    let fundamental = {
+        let mut fundamental = None;
+        for attr in tcx.get_attrs(def_id).iter() {
+            if attr.check_name("fundamental") {
+                if fundamental.is_some() {
+                    let mut err = tcx.sess.struct_span_err(
+                        attr.span,
+                        "the #[fundamental] attribute was specified at least twice");
+                    err.emit();
+                    break;
+                }
+
+                if let Some(list) = attr.meta_item_list() {
+                    if attr::list_contains_name(list, "ref_slice") {
+                        fundamental = Some(FundamentalType::SliceRef);
+                    }
+                } else {
+                    fundamental = Some(FundamentalType::All);
+                }
+            }
+        }
+        fundamental.unwrap_or(FundamentalType::None)
+    };
+
     let ty_generics = generics_of_def_id(ccx, def_id);
```

Comment:
Is there a way to report a nice error message as to why generic traits don't allow this fundamentalism? Is this something we want to pursue?